internal/build: Add contour version as a CLI option

pickledrick opened the pull request:
Fixes #2396

Signed-off-by: Peter Grant <pegrant@vmware.com>

Allows for a CLI option to print build and version information.

pickledrick commented:
Sample output:

```
> $ contour version                                                                                                                                                                                                          [±version-cmd ●]
{"branch":"version-cmd","sha":"51348913a93d32b0adae970110a850c681e51d3d","version":"v1.2.0-90-g51348913"}
```

jpeach commented on internal/build/version.go:
```diff
@@ -21,3 +32,13 @@ var Sha string
+// PrintBuildInfo prints the build information.
+func PrintBuildInfo() string {
+	buildInfo := &BuildInfo{Branch, Sha, Version}
+	out, err := json.Marshal(buildInfo)
```
Suggest that we do yaml (or manual marshaling) here?

```
$ contour version
branch: version-cmd
sha: 51348913a93d32b0adae970110a850c681e51d3d
version: v1.2.0-90-g51348913
```

There reason being that there are no other Contour commands that emit JSON, and the YAML version is easier to crack with grep, awk, etc. whereas the JSON basically needs jq.

jpeach replied:
@youngnick LMK what you think?

pickledrick replied:
Have updated to YAML

youngnick replied:
Yeah, YAML is good.

jpeach commented on cmd/contour/contour.go:
```diff
@@ -61,6 +62,7 @@ func main() {
 	sds.Arg("resources", "SDS resource filter").StringsVar(&resources)
 
 	serve, serveCtx := registerServe(app)
+	version := app.Command("version", "Build information for Contour. Contains the branch and git SHA that Contour was built from, and the Contour version.")
```
Shall we keep this short, just "Build information for Contour."

The difference being that you can just run it and see the output, which is not the case with the metric (you have to get it to serve, etc).

pickledrick replied:
Makes sense and keeps to a similar style for flags.

pickledrick replied:
Have updated to a shorter description.

pickledrick commented:
Have updated output to be YAML for readability.

```
> $ contour version                                                                                                                                                                                                          [±version-cmd ●]
branch: version-cmd
sha: 06b0d669b9531395f602e6de48081f04f50b4626
version: v1.2.0-91-g06b0d669
```

pickledrick commented:
@stevesloka given this, how about something like the following. Will have to test with make etc.

```
> $ git checkout -                                                                                                                                                                                                           [±release-1.3 ●]
Switched to branch 'version-cmd'
> $ git describe --tags --exact-match 2>/dev/null || git rev-parse --short=8 --verify HEAD                                                                                                                                   [±version-cmd ●]
ae38c692
> $ git checkout release-1.3                                                                                                                                                                                                 [±version-cmd ●]
Switched to branch 'release-1.3'
Your branch is up to date with 'upstream/release-1.3'.
> $ git describe --tags --exact-match 2>/dev/null || git rev-parse --short=8 --verify HEAD                                                                                                                                   [±release-1.3 ●]
v1.3.0
> $
```

stevesloka commented:
@pickledrick looks like that works. I just didn't think it was proper to have a non-release branch report a released version.

pickledrick commented:
> @pickledrick looks like that works. I just didn't think it was proper to have a non-release branch report a released version.

Makes sense. I have updated to this now.

youngnick commented on Makefile:
```diff
@@ -21,7 +21,7 @@ TAG_LATEST ?= false
 
 # Sets GIT_REF to a tag if it's present, otherwise the short rev.
-GIT_REF = $(shell git describe --tags || git rev-parse --short=8 --verify HEAD)
+GIT_REF = $(shell describe --tags --exact-match 2>/dev/null || git rev-parse --short=8 --verify HEAD)
```
what happened to `git describe`? `--exact-match` seems good, but I don't think just `describe` will do anything?

pickledrick replied:
my bad. Copy/Paste

jpeach commented:
> @pickledrick looks like that works. I just didn't think it was proper to have a non-release branch report a released version.

I think that the way to read `v1.3.0-24-gae38c692` is that this 24 commits passed the nearest tag, which is `v1.3.0-24` and the short hash is `ae38c692`. GitHub will actually serve the source tarball from this commit name, which surprised me.

https://git-scm.com/docs/git-describe#_examples

That said, I don't have too strong an opinion on what we want to show here when it's a non-release build.

stevesloka commented:
To me, a build from the `master` branch has no relation to v1.3.0 and is confusing. Just throwing in my 2 cents.

jpeach commented:
```
$ git describe --all --long
heads/master-0-g7696eb2
$ basename $(git describe --all --long)
master-0-g7696eb2
```

WDYT?

pickledrick commented:
> master-0-g7696eb2

I like this ^^. It's descriptive if not a tagged branch, and is quickly useful if you are not looking at a combination of branch/revision information with it.

`git describe --tags --exact-match 2>/dev/null || basename $(git describe --all --long)`

jpeach commented:
> > master-0-g7696eb2
> 
> I like this ^^. It's descriptive if not a tagged branch, and is quickly useful if you are not looking at a combination of branch/revision information with it.
> 
> `git describe --tags --exact-match 2>/dev/null || basename $(git describe --all --long)`

Leave a long block comment explaining this :)

stevesloka commented:
Why did this all change? Prev it was the git tag or the git ref if not on a tag.

jpeach commented:
> Why did this all change? Prev it was the git tag or the git ref if not on a tag.

It didn't change, it's just that now the result is visible. The build has always done this:

```
GIT_REF = $(shell git describe --tags || git rev-parse --short=8 --verify HEAD)
VERSION ?= $(GIT_REF)
```

Which gives me this result on master:
```
$ git describe --tags || git rev-parse --short=8 --verify HEAD
v1.2.0-92-gd4fb68d
```

(which is just `git describe --tags`)

jpeach commented on Makefile:
```diff
@@ -21,7 +21,7 @@ TAG_LATEST ?= false
 
 # Sets GIT_REF to a tag if it's present, otherwise the short rev.
-GIT_REF = $(shell git describe --tags || git rev-parse --short=8 --verify HEAD)
+GIT_REF = $(shell git describe --tags --exact-match 2>/dev/null || git rev-parse --short=8 --verify HEAD)
```
As per the main discussion, lets update this to:
```
git describe --tags --exact-match 2>/dev/null || basename $(git describe --all --long)
```

Please update the comment too.

jpeach commented on internal/build/version.go:
```diff
@@ -13,6 +13,17 @@
+	"gopkg.in/yaml.v2"
+)
+
+// BuildInfo is a struct for build information
```
End the comment with a full stop please :)

jpeach commented on Makefile:
```diff
@@ -20,8 +20,8 @@ TAG_LATEST ?= false
-# Sets GIT_REF to a tag if it's present, otherwise the short rev.
-GIT_REF = $(shell git describe --tags || git rev-parse --short=8 --verify HEAD)
+# Sets GIT_REF to a tag if it's present, otherwise the basename of a git describe is used.
+GIT_REF = $(shell git describe --tags --exact-match 2>/dev/null || basename $(shell git describe --all --long))
```
As per @stevesloka , let's revert the `git describe` to `git rev-parse --short=8 --verify HEAD`. Since you added the `--exact-match` flag, we will hit this case now, and the `GIT_BRANCH` variable takes care of the branch field of the long description.
